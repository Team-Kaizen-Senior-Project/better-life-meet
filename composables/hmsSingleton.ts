import type { HmsInstance } from "~/types";
let hmsInstance: HmsInstance | null = null;

export function getHmsInstance(): HmsInstance {
    if (hmsInstance === null) {
      console.log("Creating a new instance of useHms");
      hmsInstance = useHms();
    } else {
      console.log("Reusing existing instance of useHms");
    }
    return hmsInstance;
  }
