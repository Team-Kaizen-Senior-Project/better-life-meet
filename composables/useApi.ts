import type {
	Attendee,
	AttendeeFields,
	Customer,
	CustomerFields,
	Meeting,
	MeetingFields,
	Numberic,
	Pod,
	PodFields,
} from '~/types'

// All the api requests, error and response handling should happen where they are being called.

type MeetingQueryParams =
	| { podId?: never; count?: number; cursor?: string }
	| { podId: Numberic; count?: number; cursor?: never }

export const useApi = () => {
	const createCustomer = async (customer: CustomerFields) => {
		const response = await $fetch<{ data: Customer }>('/api/customer', {
			method: 'POST',
			body: customer,
		})

		const newCustomer = response.data
		return newCustomer
	}

	const getCustomer = async (id: Numberic): Promise<Customer> => {
		const response = await $fetch<{ data: Customer }>(`/api/customer/${id}`)

		const customer = response.data
		return customer
	}

	const getCustomerByEmail = async (email: string): Promise<Customer> => {
		const response = await $fetch<{ data: Customer }>(`/api/customer?email=${email}`)

		const customer = response.data
		return customer
	}

	const updateCustomer = async (id: Numberic, customer: CustomerFields): Promise<Customer> => {
		const response = await $fetch<{ data: Customer }>(`/api/customer/${id}`, {
			method: 'PATCH',
			body: customer,
		})

		const updatedCustomer = response.data
		return updatedCustomer
	}

	const deleteCustomer = async (id: Numberic): Promise<unknown> => {
		const response = await $fetch(`/api/customer/${id}`, { method: 'DELETE' })

		return response
	}

	const createPod = async (pod: PodFields): Promise<Pod> => {
		const response = await $fetch<{ data: Pod }>('/api/pod', {
			method: 'POST',
			body: pod,
		})

		const newPod = response.data
		return newPod
	}

	const getPod = async (id: Numberic): Promise<Pod> => {
		const response = await $fetch<{ data: Pod }>(`/api/pod/${id}`)

		const pod = response.data
		return pod
	}

	const updatePod = async (id: Numberic, pod: PodFields): Promise<Pod> => {
		const response = await $fetch<{ data: Pod }>(`/api/pod/${id}`, {
			method: 'PATCH',
			body: pod,
		})

		const updatedPod = response.data
		return updatedPod
	}

	const deletePod = async (id: Numberic): Promise<unknown> => {
		const response = await $fetch(`/api/pod/${id}`, { method: 'DELETE' })

		return response
	}

	const createMeeting = async (meeting: MeetingFields): Promise<Meeting> => {
		console.log(meeting)
		const response = await $fetch<{ data: Meeting }>('/api/meeting', {
			method: 'POST',
			body: meeting,
		})

		const newMeeting = response.data
		return newMeeting
	}

	const getMeeting = async (id: Numberic): Promise<Meeting> => {
		const response = await $fetch<{ data: Meeting }>(`/api/meeting/${id}`)

		const meeting = response.data
		return meeting
	}

	const getMeetings = async (params?: MeetingQueryParams): Promise<Meeting[]> => {
		const response = await $fetch<{ data: { data: Meeting[] } }>(`/api/meeting`, { params })
		//console.log(response.data.data)
		// filter out meetings that have already ended
		// sort the meetings by start time so newest is first
		return response.data.data
			.filter((meeting) => typeof meeting.endTime != 'string' && new Date(meeting.endTime.isoString) > new Date())
			.sort((a, b) => a.startTime.isoString.localeCompare(b.startTime.isoString))
	}

	const updateMeeting = async (id: Numberic, meeting: MeetingFields): Promise<Meeting> => {
		const response = await $fetch<{ data: Meeting }>(`/api/meeting/${id}`, {
			method: 'PATCH',
			body: meeting,
		})

		const updatedMeeting = response.data
		return updatedMeeting
	}

	const deleteMeeting = async (id: Numberic): Promise<unknown> => {
		const response = await $fetch(`/api/meeting/${id}`, { method: 'DELETE' })

		return response
	}

	const createAttendee = async (attendee: AttendeeFields): Promise<Attendee> => {
		const response = await $fetch<{ data: Attendee }>('/api/attendee', {
			method: 'POST',
			body: attendee,
		})

		const newAttendee = response.data
		return newAttendee
	}

	const getAttendee = async (id: Numberic): Promise<Attendee> => {
		const response = await $fetch<{ data: Attendee }>(`/api/attendee/${id}`)

		const attendee = response.data
		return attendee
	}

	const updateAttendee = async (id: Numberic, attendee: AttendeeFields): Promise<Attendee> => {
		const response = await $fetch<{ data: Attendee }>(`/api/attendee/${id}`, {
			method: 'PATCH',
			body: attendee,
		})

		const updatedAttendee = response.data
		return updatedAttendee
	}

	const deleteAttendee = async (id: Numberic): Promise<unknown> => {
		const response = await $fetch(`/api/attendee/${id}`, { method: 'DELETE' })

		return response
	}

	return {
		// Customer
		createCustomer,
		getCustomer,
		getCustomerByEmail,
		updateCustomer,
		deleteCustomer,

		// Pod
		createPod,
		getPod,
		updatePod,
		deletePod,

		// MEETINGS
		getMeetings,
		deleteMeeting,
		createMeeting,
		getMeeting,
		updateMeeting,

		// Attendee
		createAttendee,
		getAttendee,
		updateAttendee,
		deleteAttendee,
	}
}
