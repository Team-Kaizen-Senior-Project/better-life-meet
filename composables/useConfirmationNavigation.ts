export default function useConfirmNavigation(message = 'Are you sure you want to leave this page?') {
	let isLeavingConfirmed = false // Flag to track confirmation status
	const confirmNavigation = (event) => {
		if (event.type === 'beforeunload') {
			console.log('tab closed triggered')

			// Triggered on tab close
			if (!isLeavingConfirmed) {
				if (!confirm(message)) {
					event.preventDefault()
					event.returnValue = ''
				} else {
					isLeavingConfirmed = true
				}
			}
		} else if (event.type === 'popstate') {
			// Triggered on back button
			console.log('back buttib triggered')

			if (!isLeavingConfirmed) {
				if (confirm(message)) {
					history.back()
				} else {
					event.preventDefault()
				}
			}
		}
	}

	const setupConfirmNavigation = () => {
		console.log('setting up event listeners')
		window.addEventListener('beforeunload', confirmNavigation, false)
		window.addEventListener('popstate', confirmNavigation, false)
	}

	const removeConfirmNavigation = () => {
		window.removeEventListener('beforeunload', confirmNavigation, false)
		window.removeEventListener('popstate', false)
	}

	return { setupConfirmNavigation, removeConfirmNavigation }
}
