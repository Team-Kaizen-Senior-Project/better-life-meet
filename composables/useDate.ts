import dayjs from 'dayjs'
import utc from 'dayjs/plugin/utc'
import timezone from 'dayjs/plugin/timezone'
import isBetween from 'dayjs/plugin/isBetween'
import duration from 'dayjs/plugin/duration'

export const useDate = () => {
	dayjs.extend(utc)
	dayjs.extend(timezone)
	dayjs.extend(isBetween)
	dayjs.extend(duration)

	const display = (date?: string) => {
		return dayjs(date).tz().format('MMM DD YYYY hh:mm a')
	}

	return {
		display,
		dayjs,
	}
}
