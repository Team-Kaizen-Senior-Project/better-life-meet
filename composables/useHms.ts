import { ref, onBeforeUnmount, onMounted, watch } from 'vue'
import {
	HMSReactiveStore,
	selectIsLocalAudioEnabled,
	selectIsLocalVideoEnabled,
	selectPeers,
	selectIsConnectedToRoom,
	selectVideoTrackByID,
	selectHMSMessages,
	selectBroadcastMessages,
} from '@100mslive/hms-video-store'
import type { HmsInstance, ChatMessage } from '~/types'
import { select } from '@nuxt/ui/dist/runtime/ui.config'

export function useHms(): HmsInstance {
	console.log('Initializing useHms instance')
	const hmsManager = new HMSReactiveStore()
	hmsManager.triggerOnSubscribe()
	const hmsStore = hmsManager.getStore()
	const hmsActions = hmsManager.getActions()

	const userName = ref('')
	const roomCode = ref('')
	const videoRefs = ref([])

	const isLocalAudioEnabled = ref(false)
	const isLocalVideoEnabled = ref(false)
	const isConnected = ref(false)
	const peers = ref([])
	const messages: Ref<ChatMessage[]> = ref([])
	const video = useVideoStore()

	watch(peers, (newPeers) => {
		// synchronize the video elements with the peers
		newPeers.forEach((peer, index) => {
			hmsActions.attachVideo(peer.videoTrack, videoRefs.value[index])
		})
	})

	onBeforeUnmount(() => {
		leaveRoom()
	})

	const joinRoom = async (roomCode: string, username: string) => {
		const authToken = await hmsActions.getAuthTokenByRoomCode({ roomCode })
		await hmsActions.join({
			userName: username,
			authToken,
			settings: {
				isAudioMuted: true,
				isVideoMuted: !video.cameraActive,
			},
		})
	}

	const leaveRoom = async () => {
		await hmsActions.leave()
	}

	const toggleAudio = async () => {
		console.log('Toggling audio')
		isLocalAudioEnabled.value = !isLocalAudioEnabled.value
		await hmsActions.setLocalAudioEnabled(isLocalAudioEnabled.value)
	}

	const toggleVideo = async () => {
		console.log('Toggling video')
		isLocalVideoEnabled.value = !isLocalVideoEnabled.value
		await hmsActions.setLocalVideoEnabled(isLocalVideoEnabled.value)
	}

	const sendBroadcastMessage = async (message: string) => {
		await hmsActions.sendBroadcastMessage(message)
	}

	hmsStore.subscribe((newMessages) => {
		if (newMessages.length > 0) {
			const mostRecentMessage = newMessages[newMessages.length - 1]
			messages.value.push({
				id: mostRecentMessage.id,
				content: mostRecentMessage.message,
			})
		}
	}, selectHMSMessages) //for all messages, send

	hmsStore.subscribe((val) => (isLocalAudioEnabled.value = val), selectIsLocalAudioEnabled)
	hmsStore.subscribe((val) => (isLocalVideoEnabled.value = val), selectIsLocalVideoEnabled)
	hmsStore.subscribe((val) => (isConnected.value = val), selectIsConnectedToRoom)
	hmsStore.subscribe((val) => (peers.value = val), selectPeers)

	return {
		userName,
		roomCode,
		videoRefs,
		isLocalAudioEnabled,
		isLocalVideoEnabled,
		isConnected,
		peers,
		messages,
		joinRoom,
		leaveRoom,
		toggleAudio,
		toggleVideo,
		sendBroadcastMessage,
	}
}
