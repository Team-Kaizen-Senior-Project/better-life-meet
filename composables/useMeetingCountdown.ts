interface Params {
	startTime?: MaybeRef<string | undefined>
	onMeetingStart?: () => any
}

export const useMeetingCountdown = (params: Params) => {
	const { display: displayDate, dayjs } = useDate()

	const countdown = ref<string | undefined>(undefined)
	const hasStarted = ref<boolean>(false)

	const interval = ref<NodeJS.Timeout | undefined>(undefined)

	const updateCountdown = () => {
		const unrefedTime = unref(params.startTime)

		if (!unrefedTime) {
			countdown.value = undefined
			console.log('no start time is provided')
			return
		}

		const startTime = dayjs(unrefedTime)
		const now = dayjs()

		if (now.isAfter(startTime)) {
			onTimerFinish()
			return
		}

		const duration = dayjs.duration(startTime.diff(now))
		countdown.value = `Starts in ${duration.days()}d ${duration.hours()}h ${duration.minutes()}m ${duration.seconds()}s`
		hasStarted.value = false
	}

	const runInterval = () => {
		clearInterval(interval.value)
		interval.value = setInterval(() => updateCountdown(), 1000)
		updateCountdown()
	}

	const stopInterval = () => {
		clearInterval(interval.value)
	}

	const onTimerFinish = () => {
		hasStarted.value = true
		params.onMeetingStart?.()
		stopInterval()
	}

	const renderText = computed(() => (!hasStarted.value ? countdown.value : 'Meeting Has Started'))

	onMounted(() => runInterval())
	onBeforeUnmount(() => stopInterval())

	return {
		hasStarted,
		countdown,
		renderText,
	}
}
