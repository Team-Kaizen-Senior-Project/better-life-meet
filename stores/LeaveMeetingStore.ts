export const useLeaveModalStore = defineStore('leaveMeetingModal', {
    state: () => ({
      isLeaveModalOpen: false,
    }),
    actions: {
      toggleLeaveModal() {
        this.isLeaveModalOpen = !this.isLeaveModalOpen
        console.log('Leave Meeting Modal visibility:', this.isLeaveModalOpen)
      },
    },
  })