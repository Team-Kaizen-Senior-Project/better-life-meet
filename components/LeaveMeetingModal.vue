<script setup lang="ts">
import { useHms } from '~/composables/useHms'
const { isConnected } = useHms()

const leaveMeetingModal = useLeaveModalStore()
const isLeaveModalOpen = computed(() => leaveMeetingModal.isLeaveModalOpen)
const { leaveRoom } = useHms()
const customModal = {
	overlay: {
		background: 'bg-zinc-900/90 dark:bg-gray-800/75',
	},
}

async function confirmLeave() {
	try {
		await leaveRoom()
		await navigateTo('/dashboard')
		leaveMeetingModal.toggleLeaveModal()
	} catch (error) {
		console.log(error)
	}
}
</script>

<template>
	<Button v-if="isConnected" class="rounded-lg bg-rose-600 px-3 py-2 text-sm font-medium text-white hover:bg-rose-700"
		@click="leaveMeetingModal.toggleLeaveModal()" data-testid="leave-button">
		Leave
	</Button>
	<UModal v-model="isLeaveModalOpen" class="w-[90vw] max-w-[450px]" :ui="customModal" data-testid="leave-modal">
		<div class="relative rounded bg-white p-4 shadow-lg">
			<div class="text-center">
				<p class="text-lg font-semibold">Are you sure you want to leave the meeting?</p>
			</div>
			<div class="mt-4 flex justify-center gap-4">
				<UButton type="button" variant="ghost" color="gray" @click="leaveMeetingModal.toggleLeaveModal()"
					data-testid="cancel-button">
					Cancel
				</UButton>
				<UButton type="button" color="red" @click="confirmLeave">Yes, leave meeting</UButton>
			</div>
		</div>
	</UModal>
</template>
