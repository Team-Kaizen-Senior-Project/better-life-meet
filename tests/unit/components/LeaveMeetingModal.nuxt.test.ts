import { describe, it, expect } from 'vitest'
import { mountSuspended } from '@nuxt/test-utils/runtime'
import LeaveMeetingModal from '@/components/LeaveMeetingModal.vue'
import { nextTick } from 'vue'
import { useHms } from "~/composables/useHms"
import { useLeaveModalStore } from "~/stores/LeaveMeetingStore"
import {isConnected} from "happy-dom/lib/PropertySymbol";

vi.mock('~/composables/useHms', () => ({
	useHms: () => ({
		isConnected: true,
		leaveRoom: vi.fn(),
	})
}))

vi.mock('~/stores/LeaveMeetingStore', () => ({
	useLeaveModalStore: () => ({
		isLeaveModalOpen: true,
		toggleLeaveModal: vi.fn(),
	}),
}))

describe('LeaveMeetingModal', () => {
	it('Leave button renders', async () => {
		const wrapper = await mountSuspended(LeaveMeetingModal)
		const leaveButton = wrapper.find('[data-testid="leave-button"]')
		expect(leaveButton.text()).toBe('Leave')
		expect(wrapper.find('[data-testid="leave-button"]').exists()).toBe(true)
	})

	it('opens modal correctly', async () => {
		const wrapper = await mountSuspended(LeaveMeetingModal)
		const leaveButton = wrapper.find('[data-testid="leave-button"]')
		await leaveButton.trigger('click')
		await nextTick()
		expect(wrapper.html()).toMatchSnapshot()
	})
})
