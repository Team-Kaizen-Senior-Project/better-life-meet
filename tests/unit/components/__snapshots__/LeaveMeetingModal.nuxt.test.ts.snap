// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LeaveMeetingModal > opens modal correctly 1`] = `
"<button class="inline-flex items-center shadow-sm rounded-lg bg-rose-600 px-3 py-2 text-sm font-medium text-white hover:bg-rose-700 rounded-lg bg-rose-600 px-3 py-2 text-sm font-medium text-white hover:bg-rose-700" data-testid="leave-button"> Leave </button>
<!--teleport start-->
<!--teleport end-->
<div style="position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0, 0, 0, 0); white-space: nowrap; border-width: 0px; display: none;"></div>"
`;

exports[`LeaveMeetingModal > renders correctly 1`] = `"<button class="inline-flex items-center shadow-sm rounded-lg bg-rose-600 px-3 py-2 text-sm font-medium text-white hover:bg-rose-700 rounded-lg bg-rose-600 px-3 py-2 text-sm font-medium text-white hover:bg-rose-700" data-testid="leave-button"> Leave </button>"`;
